src/main/java/com/redhat/foreman/ForemanSharedNodeCloud.java

```diff
@@ -326,26 +329,67 @@ private ForemanSharedNode provision(Label label) throws Exception {
-        for (Map.Entry<String, String> host : hosts) {
-            try {
-                if (getForemanAPI().isHostFree(host.getKey())
-                        && ((label == null && Label.parse(mapData.get(host.getKey())).isEmpty())
-                        || (label != null && label.matches(Label.parse(mapData.get(host.getKey())))))) {
-                    return host.getKey();
+            Set<Map.Entry<String, String>> hosts = mapData.entrySet();
+            for (Map.Entry<String, String> host : hosts) {
+                try {
+                    if (freeHostsList.contains(host.getKey())
+                            && ((label == null && Label.parse(mapData.get(host.getKey())).isEmpty())
+                                || (label != null && label.matches(Label.parse(mapData.get(host.getKey())))))) {
+                        return host.getKey();
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(Level.SEVERE, "Unhandled exception in getHostToReserve(): ", e);
+                    e.printStackTrace();
+                    continue;
                 }
-            } catch (Exception e) {
-                LOGGER.log(Level.SEVERE, "Unhandled exception in getHostToReserve(): ", e);
-                e.printStackTrace();
-                continue;
             }
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "Unhandled exception in getHostToReserve(): ", e);
+            e.printStackTrace();
         }
+
         return null;
     }
 
     /**
+     * Get the list of hosts available for reservation for the label. Each host must be free.
+     *
+     * @param label Label to reserve for.
+     * @return list of the names of host that are free for reservation.
+     */
+    @Nonnull
+    private List<String> getFreeHostsToReserve(Label label) {
+        final List<String> hostsList = new ArrayList<String>();
+        try {
+            final List<String> freeHostsList = getForemanAPI().getAllFreeHosts();
+            Map<String, String> mapData = getHostsMapData();
+
+            Set<Map.Entry<String, String>> hosts = mapData.entrySet();
+            for (Map.Entry<String, String> host : hosts) {
+                try {
+                    if (freeHostsList.contains(host.getKey())
+                            && ((label == null && Label.parse(mapData.get(host.getKey())).isEmpty())
+                                || (label != null && label.matches(Label.parse(mapData.get(host.getKey())))))) {
+                        hostsList.add(host.getKey());
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(Level.SEVERE, "Unhandled exception in getFreeHostsToReserve(): ", e);
+                    e.printStackTrace();
+                    continue;
+                }
+            }
+        } catch (Exception e) {
+            LOGGER.log(Level.SEVERE, "Unhandled exception in getFreeHostsToReserve(): ", e);
+            e.printStackTrace();
```

Comment:
Well, you are right. We've used it for debugging purposes, however we fixed our issues already by switching to JUL from Log4j. Removing...